File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def check(self, t: int, a: List[int], k: int, m: int) -> bool:
        #超过t的不看，每一个区间为k能否有m个
        cnt = 0
        j = 0
        for i in range(len(a)):
            if a[i] > t:
                j = i + 1
                continue
            if i - j + 1 == k:
                cnt += 1
                j = i + 1
        return cnt >= m
        
    def minDays(self, a: List[int], m: int, k: int) -> int:
        if len(a) // k < m:
            return -1
        l = 0
        r = max(a)
        while l < r:
            t = (l + r) // 2
            if self.check(t, a, k, m):
                r = t
            else:
                l = t + 1
        return l  # or return r, both are same when the loop ends

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int minDays(vector<int>& bloomDay, int m, int k) {
        int n = bloomDay.size();
        if ((long long)m * k > n) return -1;
        int l = 1, r = 1e9;
        while (l < r){
            int mid = l + r >> 1;
            if (check(mid, bloomDay, m, k)){
                r = mid;
            }else{
                l = mid+1;
            }
        }
        return r;
    }

    bool check(int x, vector<int> bloomDay, int m, int k){
        int cur = 0;            //current number of bonquet
        int n = bloomDay.size();
        int seqL = 0;
        for (int i = 0; i < n; i++){
            if (bloomDay[i] <= x){
                seqL++;
            }else{
                seqL = 0;
            }
            if (seqL == k) {
                cur++;
                seqL = 0;
            }
            if (cur >= m) return true;
        }
        return false;
    }
};